yarn-project/sequencer-client/src/sequencer/sequencer.ts

```diff
@@ -289,28 +297,59 @@ export class Sequencer extends (EventEmitter as new () => TypedEventEmitter<Sequ
     this.setState(SequencerState.PROPOSER_CHECK, slot);
     const [canPropose, proposer] = await this.checkCanPropose(slot);
 
-    // If we are not a proposer, check if we should invalidate a invalid block, and bail
+    // If we are not a proposer check if we should invalidate a invalid block, and bail
     if (!canPropose) {
       await this.considerInvalidatingBlock(syncedTo, slot);
       return;
     }
 
+    // In fisherman mode, check if we've already validated this slot to prevent duplicate attempts
+    if (this.config.fishermanMode) {
+      if (this.lastSlotForValidationBlock === slot) {
+        this.log.trace(`Already validated block building for slot ${slot} (skipping)`, { slot });
+        return;
+      }
+      this.log.debug(
+        `Building validation block for slot ${slot} (actual proposer: ${proposer?.toString() ?? 'none'})`,
+        { slot, proposer: proposer?.toString() },
+      );
+      // Mark this slot as being validated
+      this.lastSlotForValidationBlock = slot;
+    }
+
     // Check that the slot is not taken by a block already (should never happen, since only us can propose for this slot)
     if (syncedTo.block && syncedTo.block.header.getSlot() >= slot) {
       this.log.warn(
         `Cannot propose block at next L2 slot ${slot} since that slot was taken by block ${syncedTo.blockNumber}`,
         { ...syncLogData, block: syncedTo.block.header.toInspect() },
       );
+      this.metrics.recordBlockProposalPrecheckFailed('slot_already_taken');
       return;
     }
 
     // We now need to get ourselves a publisher.
     // The returned attestor will be the one we provided if we provided one.
     // Otherwise it will be a valid attestor for the returned publisher.
-    const { attestorAddress, publisher } = await this.publisherFactory.create(proposer);
+    // In fisherman mode, pass undefined to use the fisherman's own keystore instead of the actual proposer's
+    const { attestorAddress, publisher } = await this.publisherFactory.create(
+      this.config.fishermanMode ? undefined : proposer,
+    );
     this.log.verbose(`Created publisher at address ${publisher.getSenderAddress()} for attestor ${attestorAddress}`);
     this.publisher = publisher;
 
+    // In fisherman mode, set the actual proposer's address for simulations
+    if (this.config.fishermanMode) {
+      if (proposer) {
+        publisher.setProposerAddressForSimulation(proposer);
+        this.log.debug(`Set proposer address ${proposer} for simulation in fisherman mode`);
+      } else {
+        this.log.warn(
+          `Fisherman mode active but no specific proposer for slot ${slot} (anyone can propose). Simulations will use fisherman's own account which may have insufficient funds.`,
```

Comment:
Why not use a state override to assign funds to it?

Reply:
yeah we do now. Removing this log